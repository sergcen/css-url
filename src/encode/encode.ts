import { promises as asyncFs } from 'fs';
import * as mime from 'mime';
import svgToTinyDataUri from 'mini-svg-data-uri';
import FileType from 'file-type';

export type TEncodeType = 'auto' | 'base64' | 'svg' | undefined;

export const encodeSvg: (
    content: Buffer | string,
    optimized?: boolean
) => string = (content, optimized = true): string => {
    if (Buffer.isBuffer(content)) {
        content = content.toString('utf8');
    }

    return optimized
        ? /**
           * Optimize encoding SVG files (IE9+, Android 3+)
           * @see https://codepen.io/tigt/post/optimizing-svgs-in-data-uris
           */
          svgToTinyDataUri(content)
        : `data:image/svg+xml,${encodeURIComponent(content)}`;
};

export const encodeBase64: (
    content: Buffer | string,
    mime: string
) => string = (content, mime) => {
    if (typeof content === 'string') {
        content = Buffer.from(content);
    }

    const dataMime = `data:${mime};base64`;

    return `${dataMime},${content.toString('base64')}`;
};

export const encode: (
    content: Buffer | string,
    mime?: FileType.MimeType | 'image/svg+xml' | string,
    encodeType?: TEncodeType,
    optimizeSvgEncode?: boolean,
) => string = (
    content,
    mime,
    encodeType = 'auto',
    optimizeSvgEncode = true,
) => {    
    if (!mime) {
        throw Error('Mime type not recognized');
    }

    if (encodeType === 'auto') {
        encodeType = mime === 'image/svg+xml' ? 'svg' : 'base64';
    }

    if (encodeType === 'base64') {
        return encodeBase64(content, mime);
    }

    return encodeSvg(content, optimizeSvgEncode);
};

export const encodeFile: (
    filePath: string,
    encodeType?: TEncodeType,
    optimizeSvgEncode?: boolean
) => Promise<string> = async (filePath, encodeType, optimizeSvgEncode) => {
    let fileMime: string | null | undefined = mime.getType(filePath);
    const fileContentBuffer = await asyncFs.readFile(filePath);

    if (!fileMime && Buffer.isBuffer(fileContentBuffer)) {
        fileMime = (await FileType.fromBuffer(fileContentBuffer))?.mime;
    }

    if (!fileMime) {
        throw Error(`Couldn\'t parse mime type from ${filePath}`);
    }


    return encode(fileContentBuffer, fileMime, encodeType, optimizeSvgEncode);
};
