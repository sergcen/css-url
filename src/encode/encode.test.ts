import * as fs from 'fs';
import { encodeFile, encode } from './encode';
import * as path from 'path';

const fixturePath = (name: string) => {
    return path.join(__dirname, 'fixtures', name);
}

const SVG = fixturePath('encode-sample.svg');
const PNG = fixturePath('encode-sample.png');
const TXT = fixturePath('encode-text.txt');

describe('parse css url', () => {
    it('encode file', async () => {
        expect(await encodeFile(SVG)).toMatchSnapshot();
        expect(await encodeFile(PNG)).toMatchSnapshot();
    });

    it('encode svg without optimization', async () => {
        expect(await encodeFile(SVG, 'auto', false)).toMatchSnapshot();
    });

    it('encode', async () => {
        const fileBuffer = await fs.promises.readFile(PNG);

        expect(await encode(fileBuffer, 'image/png')).toMatchSnapshot();
    });

    it('encode (string content)', async () => {
        const fileStr = await fs.promises.readFile(TXT, 'utf-8');
        const fileBuffer = await fs.promises.readFile(TXT);

        expect(await encode(fileStr, 'image/png')).toEqual(await encode(fileBuffer, 'image/png'))
    });
});
