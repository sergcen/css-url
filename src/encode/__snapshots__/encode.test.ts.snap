// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse css url encode 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQAAAAA3bvkkAAAAAnRSTlMAAHaTzTgAAAAKSURBVHgBY2AAAAACAAFzdQEYAAAAAElFTkSuQmCC"`;

exports[`parse css url encode file 1`] = `"data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 100 100'%3e %3cpath d='M10%2c10 H90 L50%2c70'/%3e%3ctext y='90'%3e'' %26apos%3b %23 %25 %26amp%3b %c2%bf %f0%9f%94%a3%3c/text%3e %3c/svg%3e"`;

exports[`parse css url encode file 2`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABAQAAAAA3bvkkAAAAAnRSTlMAAHaTzTgAAAAKSURBVHgBY2AAAAACAAFzdQEYAAAAAElFTkSuQmCC"`;

exports[`parse css url encode svg without optimization 1`] = `"data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20100%20100%22%3E%0A%0A%20%20%20%20%3Cpath%20d%3D%22M10%2C10%20H90%20L50%2C70%22%2F%3E%3Ctext%20y%3D%2290%22%3E%22'%20%26apos%3B%20%23%20%25%20%26amp%3B%20%C2%BF%20%F0%9F%94%A3%3C%2Ftext%3E%0A%3C%2Fsvg%3E%0A"`;
